### ISSUE TITLE
Can't use pipes when reading from STDIN

### ISSUE THREAD

**Issue by slundberg** [JuliaLang/julia#8529]
I was trying to stream data into a Julia script that used run() to execute a chain of external programs. This broke and I distilled the problem to the following example. This is with the latest stable version (Version 0.3.2-pre+33 (2014-09-29 19:58 UTC)) and is reproducible on Debian and Mac OS X.

When given the following simple script we will call `test.jl`:

```julia
f = length(ARGS) > 0 ? open(ARGS[1]) : STDIN
for line in eachline(f)
    run(`echo asdf` |> `cat`)
end
```

The following works as expected:

```
> echo test > some_text_file
> julia test.jl some_text_file
asdf
```

as does:

```
> julia test.jl < some_text_file
asdf
```

but it fails when piping to STDIN:

```
> cat some_text_file | julia test.jl
ERROR: pipes passed to spawn must be initialized
 in spawn at process.jl:301
 in spawn at process.jl:259
 in spawn at process.jl:389
 in anonymous at no file:3
 in include at ./boot.jl:245
 in include_from_node1 at loading.jl:128
 in process_options at ./client.jl:285
 in _start at ./client.jl:354
 in _start at /Users/slund1/projects/julia/usr/lib/julia/sys.dylib
while loading /Users/slund1/Desktop/test.jl, in expression starting on line 2
```

**Comment by bicycle1885** [comment]
This problem is inevitable when a pipe is used together with the `download` function.
I'm not sure why this happens, but the handle of a pipe oddly starts to point to the null pointer just after reading a byte from it.

```julia
~/v/julia (master|…) $ cat bug.jl
@show typeof(STDIN)
@show STDIN.handle
@show read(STDIN, UInt8)
@show STDIN.handle
~/v/julia (master|…) $ echo "foo" | ./julia bug.jl
typeof(STDIN) = Base.Pipe
STDIN.handle = Ptr{Void} @0x00007fb3a9f009b0
read(STDIN,UInt8) = 0x66
STDIN.handle = Ptr{Void} @0x0000000000000000
```

**Comment by bicycle1885** [comment]
@Keno 
Could you check my pull request to fix the problem? #10469
